### ISSUE TITLE
Clarification: Is this Multi-Threaded or Not?

### ISSUE THREAD

**Issue by solsnare** [Cysharp/UniTask#317]
Not specifically an issue, but didn't know where else to put this.

I see on the main page that it doesn't use multi-threading

`Runs completely on Unity's PlayerLoop so doesn't use threads and runs on WebGL, wasm, etc.`

But further down in the code samples, it seems to make mention of throwing process into a "multi-threaded thread pool"

```
// Multithreading, run on ThreadPool under this code
await UniTask.SwitchToThreadPool();
```

The reason I ask about this clarification is because I am currently trying to validate if another source here is actually multi-threading or not, because performance is really important and if it's not actually multi-threading, I'm worried that could be causing the issue.

More specifically, this is the project I'm performance testing:
[https://github.com/asus4/tf-lite-unity-sample](https://github.com/asus4/tf-lite-unity-sample)

And even more specifically, the file in particular is trying to run a BlazePose Pose Detection algorithm:
[https://github.com/asus4/tf-lite-unity-sample/blob/master/Assets/Samples/BlazePose/BlazePoseSample.cs](https://github.com/asus4/tf-lite-unity-sample/blob/master/Assets/Samples/BlazePose/BlazePoseSample.cs)

The code that executes at line 240 is run when you make it "run in background" which seemingly releases the Unity main thread of the burden of running the Pose Detection code, and it does a great job of that.. However, the performance statistics of the Pose detection frames coming in are HALF of what it was when it wasn't threaded in the background... This is what made me wonder if instead of it being on it's own thread/core running at the maximum possible speed, if it's being throttled by spreading it's self over multiple frames, similar to running it async in a coroutine like structure.

Trying to port the code to be in the C# threading scheme to validate if this is the case or not is proving to be difficult and time consuming due to issues of not being allowed to run certain things off the main thread, so it would be ideal to validate if this is in fact running in a separate thread or not.

I tried using the profiler's timeline feature to validate if it was using a separate thread or not, and it seems to be saying that it is, since those pink lines don't show up in the timeline if background threading is off:

![image](https://user-images.githubusercontent.com/6552072/133179676-5659d805-7ecd-4c32-b841-3a5fc1202092.png)

TL;DR

I need to know if this plugin actually uses real multi-threading that can run on separate hardware threads/cores, or if the "multiple-threads" are all running on the main thread as is written in the feature description of this plugin.

Further question if it is in-fact using multi-threading in the thread pool, why is the performance lower than if the code was running in the main thread?  does it have something to do with the PlayerLoopTiming variable ? I tried changing it to Update instead of FixedUpdate but that didn't do anything notable for the performance.

Clarification would be much appreciated, thank you!

**Comment by neuecc** [comment]
You are correct, it is without SwitchToThreadPool.
SwitchToThreadPool sends the execution code to ThreadPool.QueueUserWorkItem.

Isn't the cause of the performance degradation the frequency of SwitchToMainThread usage?
To maximize performance, put the executable code on the ThreadPool and do switch to mainthread only once at the last timing.
If you call SwitchToMainThread many times, the performance will degrade by the waiting time of MainThraed.

In code, poseDetect.InvokeAsync and poseLandmark.InvokeAsync both have SwitchToMainThread.
In other words, it will be twice as slow because it waits for the MainThread twice.

**Comment by solsnare** [comment]
This clarification was exactly what I needed, and you found the exact bug that I'm dealing with, should be smooth sailing now.

Can't thank you enough!
